### PR TITLE
Add initial authentication storage

### PR DESCRIPTION
**Pull request by nielsvanvelzen**
**Changes**

- Store credentials
  - Uses Android's AccountManager to store access tokens and the ability to remove accounts from within the system settings
- Display all servers from: storage, server discovery, legacy credentials
- Display all users from: public user endpoint, storage, legacy credentials
- Display errors when adding a server/user (rough implementation)
- Sign in
- Backup Jellyfin app settings and credentials (without access tokens)

**Future pull requests**

- Cache & show profile pictures
- Remember last user

**TODO**

- [x] Remove accidently committed dev note files
- [ ] Squash all commits
- [x] Rewrite legacy credential migration code
- [x] Clean up all code

**Issues**
Part of #612 

**Dev notes**

The AlertFragment uses callbacks passed via the constructor. This is not a good idea for 2 reasons
1) Fragment's should not have custom constructors per Android guidelines, [this StackOverflow article](https://stackoverflow.com/questions/10450348/do-fragments-really-need-an-empty-constructor) explains it a bit.
2) In this case the callbacks came from another fragment which was unloaded at the time the callback executed. This caused issues were the viewmodel was unavailable and crashed the app.

We should aim for fragments that are not dependent on other fragments/activies. All UI should only interact via viewmodels with repositories/services/etc. and data should be passed via Bundles.

**Other**

I decided to use UUID everywhere since this is what the new apiclient does too. In order to make things work I added a few helper methods to convert string -> uuid that support the "simple" format (which doesn't use hyphens). Because those UUID's differ from the strings in the old credentials I needed to convert strings -> uuid -> string a bunch of times. This will all be solved when we remove the legacy credential support and upgrade to the new apiclient.

I've also decided to use Kotlin's flow internally and convert it to Android's LiveData at the end of the pipeline because a flow is much easier to work with and provides a more rich API. I added a helper method for flows too to collect them to a `LiveData<List<T>>`.

### CODE REVIEW

**Comment by nielsvanvelzen** [comment]
This PR became quite massive and I'm a bit lost in what still needs to happen etc. so I did some final cleaning and marking it ready or review.

I think there's still some small bugs which I'd say are easier to fix in future PR's instead of this one, but let me know what you think about that @thornbill.

**Comment by thornbill** [comment]
This is crashing on launch for me on a Fire Stick:

```
java.lang.IllegalStateException: Flow invariant is violated:
		Flow was collected in [StandaloneCoroutine{Active}@c69d5fa, Main [immediate]],
		but emission happened in [DispatchedCoroutine{Active}@19e77ab, LimitingDispatcher@32313e08[dispatcher = DefaultDispatcher]].
		Please refer to 'flow' documentation or use 'flowOn' instead
	at kotlinx.coroutines.flow.internal.SafeCollector_commonKt.checkContext(SafeCollector.common.kt:84)
	at kotlinx.coroutines.flow.internal.SafeCollector.checkContext(SafeCollector.kt:82)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:68)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:54)
	at org.jellyfin.androidtv.auth.ServerRepositoryImpl$getDiscoveryServers$1$1$invokeSuspend$$inlined$map$1$2.emit(Collect.kt:139)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:14)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:71)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:54)
	at org.jellyfin.apiclient.discovery.ServerDiscovery$discover$1.invokeSuspend(ServerDiscovery.kt:111)
	at org.jellyfin.apiclient.discovery.ServerDiscovery$discover$1.invoke(ServerDiscovery.kt)
	at kotlinx.coroutines.flow.SafeFlow.collect(Builders.kt:56)
	at org.jellyfin.androidtv.auth.ServerRepositoryImpl$getDiscoveryServers$1$1$invokeSuspend$$inlined$map$1.collect(SafeCollector.common.kt:114)
	at org.jellyfin.androidtv.auth.ServerRepositoryImpl$getDiscoveryServers$1$1.invokeSuspend(ServerRepository.kt:104)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

**Comment by nielsvanvelzen** [comment]
Discovery crash should be fixed now.

Used [this](https://elizarov.medium.com/execution-context-of-kotlin-flows-b8c151c9309b) source (which explains flows and coroutines quite well). Basically I did an illegal operation by using `withContext` in the flow. I should've applied the scope to the flow by using `flowOn` instead.

**Comment by nielsvanvelzen** [comment]
Bug found:
- Login using auto discovery doesn't work - when authenticating the server does not exists yet and thus the address can not be found.
  - Fix for now is to manually add the server

**Comment by thornbill** [comment]
I've found a couple things so far:

1. Discovered server is displayed before manually added server
2. Saved user listed twice if it's also a saved user
3. Passwordless login no longer bypasses the login form

File: app/src/main/java/org/jellyfin/androidtv/auth/AuthenticationRepository.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+			accessToken = user.accessToken
+		})
+
+		try {
+			val userDto = callApi<UserDto?> { callback -> apiClient.GetUserAsync(user.id.toString(), callback) }
+			if (userDto != null) {
+				application.currentUser = userDto
+				return true
+			}
+		} catch (err: Exception) {
```
Comment (thornbill) [inline app/src/main/java/org/jellyfin/androidtv/auth/AuthenticationRepository.kt:69]: We should suppress these warnings.

**Comment by thornbill** [comment]
The tests are failing due to a reference to some old code

```
e: /home/vsts/work/1/s/app/src/test/java/org/jellyfin/androidtv/util/apiclient/ModelExtensionsTest.kt: (45, 5): Unresolved reference: toUser
```

**Comment by nielsvanvelzen** [comment]
I realized that I accidently pushed some changes I was working on for the CardPresenter to optimize it. But since those changes were ready for review I'll leave them in.

**Comment by thornbill** [comment]
Just tested this again in an emulator and noticed the following issues:
1. Passwordless login of public users no longer bypasses the login form
2. Adding multiple servers seems to not work only my saved server is ever displayed.

**Comment by nielsvanvelzen** [comment]
> Passwordless login of public users no longer bypasses the login form

I just realized it only works for users not saved. I'll fix that.

> Adding multiple servers seems to not work only my saved server is ever displayed.

Did you restart the app? There is a bug where new servers only show up after restart. The fix for that one will be in another PR because I expect it to be quite large.

**Comment by nielsvanvelzen** [comment]
_copied from commit message_

We now remember if a user has a password (using a boolean) in the offline data set. If a user fails to authenticate because a password was added it will prompt for a password. If it's the other way around the user can just proceed in the password-prompt with an empty password and the boolean will be updated for future logins.

**Comment by thornbill** [comment]
I still seem to be randomly prompted for a password when logging in and out with a passwordless user. This can probably be addressed later though. Let's get this in once the conflicts are resolved. :+1:

**Review by thornbill — APPROVED** [review]

**Review by thornbill — APPROVED** [review]

**Pull request merged**